File: javascript/bob/bob.js
export const hey = message => {
  const remark = message.trim()

  if (isSilence(remark)) {
    return "Fine. Be that way!"
  } else {
    return respondToVerbalRemark(remark)
  }
}

function isSilence(remark) {
  return remark === ""
}

function respondToVerbalRemark(remark) {
  const question = isQuestion(remark)
  const shouting = isShouting(remark)

  if (question && shouting) {
    return "Calm down, I know what I'm doing!"
  } else if (question) {
    return "Sure."
  } else if (shouting) {
    return "Whoa, chill out!"
  } else {
    return "Whatever."
  }
}

function isQuestion(remark) {
  return remark.endsWith("?")
}

function isShouting(remark) {
  return hasLetters(remark) && remark.toUpperCase() === remark
}

function hasLetters(remark) {
  return LETTERS.test(remark)
}

const LETTERS = /[A-Za-z]/i

File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1]).


response(String) ->
  case string:trim(String) of
    "" ->
      "Fine. Be that way!";

    Remark ->
      respond_to_verbal_remark(Remark)
  end.

% Private

respond_to_verbal_remark(Remark) ->
  case {is_question(Remark), is_yelling(Remark)} of
    {true, true} ->
      "Calm down, I know what I'm doing!";

    {true, false} ->
      "Sure.";

    {false, true} ->
      "Whoa, chill out!";

    {false, false} ->
      "Whatever."
  end.

is_question(Remark) ->
  lists:last(Remark) =:= $?.

is_yelling(Remark) ->
  Remark =:= string:uppercase(Remark) andalso
    Remark =/= string:lowercase(Remark).
